File: app/src/main/java/se/gustavkarlsson/aurora_notifier/android/background/providers/impl/aggregating_aurora_factors/GetVisibility.java
package se.gustavkarlsson.aurora_notifier.android.background.providers.impl.aggregating_aurora_factors;

import android.location.Location;
import android.util.Log;

import java.util.concurrent.Callable;
import java.util.concurrent.TimeoutException;

import se.gustavkarlsson.aurora_notifier.android.background.providers.VisibilityProvider;
import se.gustavkarlsson.aurora_notifier.android.models.factors.Visibility;

class GetVisibility implements ErrorHandlingTask<Visibility> {
	private static final String TAG = GetVisibility.class.getSimpleName();

	private final VisibilityProvider provider;
	private final Location location;

	GetVisibility(VisibilityProvider provider, Location location) {
		this.provider = provider;
		this.location = location;
	}

	@Override
	public Callable<Visibility> getCallable() {
		return this::call;
	}

	private Visibility call() throws Exception {
		Log.i(TAG, "Getting visibility...");
		Visibility visibility = provider.getVisibility(location.getLatitude(), location.getLongitude());
		Log.d(TAG, "Visibility is:  " + visibility);
		return visibility;
	}

	@Override
	public Visibility handleInterruptedException(InterruptedException e) {
		return new Visibility();
	}

	@Override
	public Visibility handleThrowable(Throwable e) {
		return new Visibility();
	}

	@Override
	public Visibility handleTimeoutException(TimeoutException e) {
		return new Visibility();
	}
}

File: app/src/main/kotlin/se/gustavkarlsson/skylight/android/background/providers/impl/aggregating_aurora_factors/GetVisibility.kt
package se.gustavkarlsson.skylight.android.background.providers.impl.aggregating_aurora_factors

import android.location.Location
import org.jetbrains.anko.AnkoLogger
import org.jetbrains.anko.debug
import org.jetbrains.anko.info
import se.gustavkarlsson.skylight.android.background.providers.VisibilityProvider
import se.gustavkarlsson.skylight.android.models.Visibility
import java.util.concurrent.Callable
import java.util.concurrent.TimeoutException

class GetVisibility(
		private val provider: VisibilityProvider,
		private val location: Location
) : ErrorHandlingTask<Visibility>, AnkoLogger {

    override val callable: Callable<Visibility>
        get() = Callable { this.call() }

    private fun call(): Visibility {
        info("Getting visibility...")
        val visibility = provider.getVisibility(location.latitude, location.longitude)
        debug("Visibility is:  $visibility")
        return visibility
    }

    override fun handleInterruptedException(e: InterruptedException) = Visibility()

    override fun handleThrowable(e: Throwable) = Visibility()

    override fun handleTimeoutException(e: TimeoutException) = Visibility()
}
